src/overlay/FlowControlCapacity.cpp

```diff
@@ -29,24 +29,74 @@ FlowControlMessageCapacity::getMsgResourceCount(StellarMessage const& msg) const
 FlowControlCapacity::ReadingCapacity
 FlowControlMessageCapacity::getCapacityLimits() const
 {
-    return {mApp.getConfig().PEER_FLOOD_READING_CAPACITY,
-            mApp.getConfig().PEER_READING_CAPACITY};
+    return {
+        mApp.getConfig().PEER_FLOOD_READING_CAPACITY,
+        std::make_optional<uint64_t>(mApp.getConfig().PEER_READING_CAPACITY)};
 }
 
 void
 FlowControlMessageCapacity::releaseOutboundCapacity(StellarMessage const& msg)
 {
     ZoneScoped;
-    releaseAssert(msg.type() == SEND_MORE);
+    releaseAssert(msg.type() == SEND_MORE || msg.type() == SEND_MORE_EXTENDED);
     auto numMessages = FlowControl::getNumMessages(msg);
     if (!hasOutboundCapacity(msg) && numMessages != 0)
     {
-        CLOG_DEBUG(Overlay, "Got outbound capacity for peer {}",
+        CLOG_DEBUG(Overlay, "Got outbound message capacity for peer {}",
                    mApp.getConfig().toShortString(mNodeID));
     }
     mOutboundCapacity += numMessages;
 }
 
+bool
+FlowControlMessageCapacity::canRead() const
+{
+    ZoneScoped;
+    releaseAssert(mCapacity.mTotalCapacity);
+    return *mCapacity.mTotalCapacity > 0;
+}
+
+FlowControlByteCapacity::FlowControlByteCapacity(Application& app,
+                                                 NodeID const& nodeID)
+    : FlowControlCapacity(app, nodeID)
+{
+    mCapacity = getCapacityLimits();
+}
+
+FlowControlCapacity::ReadingCapacity
+FlowControlByteCapacity::getCapacityLimits() const
```

Comment:
Just curious, why do we have a message count reading capacity but not a byte based reading capacity?

Reply:
good question. the reading capacity is a purely internal concept, needed to ensure fairness across all of our connections (what we used to do was read from a peer while there was any data available, potentially starving other peers). the reason I decided to keep only the message counting for now was because the largest message we allow in overlay is 16MB anyway, so it seemed like it wasn't super useful to try to enforce the existing byte limits, since they would be way off when we receive larger messages like tx sets. That being said, we might do something better than this in the future, but for now I'm leaving it as-is.